test/Autofac.Extras.DynamicProxy.Test/ServiceMiddlewareInterfaceInterceptorsFixture.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Copyright (c) Autofac Project. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using Castle.DynamicProxy;
+using Moq;
+using IInvocation = Castle.DynamicProxy.IInvocation;
+
+namespace Autofac.Extras.DynamicProxy.Test;
+
+public class ServiceMiddlewareInterfaceInterceptorsFixture
+{
+    public interface IPublicInterface
+    {
+        string PublicMethod();
+    }
+
+    public interface IDecoratorInterface
+    {
+        void Decorate();
+    }
+
+    [Fact]
+    public void InterceptsPublicInterfacesUseGenericMethod()
+    {
+        Mock<IDecoratorInterface> mockDecorator = new();
+        ContainerBuilder builder = new();
+        builder.RegisterType<StringMethodInterceptor>();
+        builder.RegisterDecorator<Decorator, IPublicInterface>();
+        builder.RegisterInstance(mockDecorator.Object);
+        builder
+            .RegisterType<Interceptable>()
+            .InterceptedBy(typeof(StringMethodInterceptor))
+            .As<IPublicInterface>();
+        builder.EnableInterfaceInterceptors<IPublicInterface>();
```

Comment:
I find this syntax to be confusing. We already registered a type to have interceptors, but then there's _another_ line to indicate interception.

I feel like it should be maybe more like...

```c#
builder.RegisterDecorator<T, U>().InterceptedBy(typeof(Z));
```

Something analogous to the current syntax.

Reply:
> I find this syntax to be confusing. We already registered a type to have interceptors, but then there's _another_ line to indicate interception.
> 
> I feel like it should be maybe more like...
> 
> ```cs
> builder.RegisterDecorator<T, U>().InterceptedBy(typeof(Z));
> ```
> 
> Something analogous to the current syntax.

If I change it to all in one line Is it ok? like this,

```
builder.RegisterType<Interceptable>().InterceptedBy(typeof(StringMethodInterceptor)).As<IPublicInterface>();
```

but I saw other file like InterfaceInterceptionWithPropertyInjectionFixture.cs

use 
```
 builder
            .RegisterType<InterceptableWithProperty>()
            .PropertiesAutowired()
            .EnableInterfaceInterceptors()
            .InterceptedBy(typeof(StringMethodInterceptor))
            .As<IPublicInterface>();
```
should we keep same style?

Reply:
The thing that makes the new syntax weird is that it's not connected to a RegisterType or RegisterDecorator. I think if it reads like a sentence - "register X as Y intercepted by Z" - that would be consistent and easier to understand.

Reply:
This is the bit currently pending. I'm not sure we want a syntax where it's some sort of builder level registration. I find the syntax to be challenging because there are now two ways to register things. It seems like if the interception needs to apply to a decorator, it should be attached to the decorator registration.

Reply:
Sorry, I've only had my phone, away for the holiday season, and I'm trying to answer quickly to help but also haven't been able to really look at anything on a big monitor and provide robust feedback.

In the test highlighted here, I see the new interception syntax like this:

```c#
// Register the decorator.
builder.RegisterDecorator<Decorator, IPublicInterface>();

// Register a type that needs to be intercepted.
builder
  .RegisterType<Interceptable>()
  .EnableInterfaceInterceptors()
  .InterceptedBy(typeof(StringMethodInterceptor))
  .As<IPublicInterface>();

// This is the confusing/inconsistent part - this standalone interception line.
builder.EnableInterfaceInterceptors<IPublicInterface>();
```

The thing that's inconsistent here is that, generally speaking, the syntax for registering things is all built up and attached to one registration.

```c#
// Start a registration
builder

  // Register a type `Interceptable`
  .RegisterType<Interceptable>()

  // I want to enable interface interception for this type
  .EnableInterfaceInterceptors()

  // Say what that type is intercepted by
  .InterceptedBy(typeof(StringMethodInterceptor))

  // Expose the type as `IPublicInterface`
  .As<IPublicInterface>();
```

It's a chained builder style syntax. I know that when I register the interceptor it will be intercepting things for `RegisterType`. I know the `As` applies to that registration.

The new line here:

```c#
builder.EnableInterfaceInterceptors<IPublicInterface>();
```

This is inconsistent. It reads like it's enabling some sort of _global_ interception, but if I read that in context of the other syntax...

- Why would I need `.EnableInterfaceInterceptors()` on the `RegisterType<T>` call if I'm supposed to use this new standalone `builder.EnableInterfaceInterceptors<T>()` syntax?
- What happens if I drop `.EnableInterfaceInterceptors()` on the `RegisterType<T>` call and _only_ use the new standalone syntax?
- I'm enabling interface interception, but by _what_? What's intercepting it? How do I control that?

I know that you can probably answer that here in comments, but that's not the point - the point is that someone later _reading the code_ or _trying to figure out how to make interception work_ is going to have these questions. The syntax shouldn't be confusing like that or cause folks to wonder what it means.

That's what I'm stuck on here.

If the goal of this PR is to allow interface interception on decorators, I would expect to see something more like:

```c#
[Fact]
public void InterceptsPublicInterfacesUseGenericMethod()
{
  ContainerBuilder builder = new();
  builder.RegisterType<StringMethodInterceptor>();

  // Same support for adding interception to a decorator
  // that we have for existing registrations. Familiar, consistent
  // syntax.
  builder
    .RegisterDecorator<Decorator, IPublicInterface>()
    .EnableInterfaceInterceptors()
    .InterceptedBy(typeof(StringMethodInterceptor);

  // Same syntax as before.
  builder
    .RegisterType<Interceptable>()
    .EnableInterfaceInterceptors()
    .InterceptedBy(typeof(StringMethodInterceptor))
    .As<IPublicInterface>();

  // ... and the rest of the test
}
```

Perhaps a good test to verify the syntax works is to see if we can get the _decorator_ to be intercepted by a _different interceptor_ than the thing being decorated.

```c#
[Fact]
public void DifferentInterceptorsForEachComponent()
{
  ContainerBuilder builder = new();
  builder.RegisterType<StringMethodInterceptor>();
  builder.RegisterType<CallCountInterceptor>();

  // Decorator gets intercepted by one thing
  builder
    .RegisterDecorator<Decorator, IPublicInterface>()
    .EnableInterfaceInterceptors()
    .InterceptedBy(typeof(CallCountInterceptor);

  // Base service gets intercepted by something else
  builder
    .RegisterType<Interceptable>()
    .EnableInterfaceInterceptors()
    .InterceptedBy(typeof(StringMethodInterceptor))
    .As<IPublicInterface>();

  // ... and the rest of the test
}
```

There are other features, too, that need to work the same if we add this.

- [Being able to add proxy generation options when enabling interception for the decorator.](https://github.com/autofac/Autofac.Extras.DynamicProxy/blob/2115f37e00614fa0507ad5dae984622558a6dc0f/test/Autofac.Extras.DynamicProxy.Test/InterceptorsChosenByMetadataWithOptionsFixture.cs#L19)
- [Being able to mark the decorator with an attribute to select the interceptor.](https://github.com/autofac/Autofac.Extras.DynamicProxy/blob/2115f37e00614fa0507ad5dae984622558a6dc0f/test/Autofac.Extras.DynamicProxy.Test/AttributedInterfaceInterceptionFixture.cs#L11)
- Possibly needing to differentiate between class and interface interceptors - can the class _being decorated_ have `EnableClassInterceptors` but the _decorator_ has `EnableInterfaceInterceptors`? What about the other way around?

That's the syntax and functionality that I'm stuck on, and it's why the standalone `builder.EnableInterfaceInterceptors<T>()` doesn't really work in the ecosystem.